### ISSUE TITLE
gsutil -m fails on system with many threads

### ISSUE THREAD

**Issue by afg1** [GoogleCloudPlatform/gsutil#283]
When running `gsutil -m  cp -r gs://some/path/* ./some/other/path/` on a laptop (4 threads) everything is fine, but when running an identical command on a server (48 threads), gsutil crashes.

Things seem to run okay for a while, but then an error like: 
```
Your "Oauth 2.0 User Account" credentials are invalid. Please run
$ gcloud auth login
```
appears, despite the login having been okay for a while. 

I tested this on a recent python (2.7.10) and a much older version (2.7 r27:82500) with the same result on both. 

Output of `gsutil version -l` which I hear can be useful:

```
gsutil version: 4.13
checksum: 141a3e09b42e1b0b6033108aa24c2286 (OK)
boto version: 2.38.0
python version: 2.7.10 (default, Jul  2 2015, 15:37:21) [GCC 5.1.0]
OS: Linux 2.6.35.13-92.fc14.x86_64
multiprocessing available: True
using cloud sdk: True
config path: /home/agreen/.config/gcloud/legacy_credentials/a.f.green1@gmail.com/.boto
gsutil path: /home/agreen/local/google-cloud-sdk/platform/gsutil/gsutil
compiled crcmod: False
installed via package manager: False
editable install: False
```

**Comment by thobrla** [comment]
Do you know if a service account was ever used on this server?  I wonder if https://github.com/GoogleCloudPlatform/gsutil/commit/98caaff874abeb097f222f1fb00e42d5089961bf may fix this.

When you say "48 threads", do you mean you've set parallel_thread_count to 48 in your .boto config?  If so, what is parallel_process_count?

**Comment by afg1** [comment]
I'm running this on a local machine, so it is using my account, not a service account.

As I know I don't have a .boto config file. I can try creating one and setting parallel_thread_count if that would be helpful?

**Comment by thobrla** [comment]
If you're not configuring threads in a .boto file (gcloud creates one for you, as noted above), how do you know that in one case you have 4 threads, and in another, 48?  Are you running 48 separate instances of gsutil?

**Comment by afg1** [comment]
Ah, sorry, I didn't read my own debug output...

Comparing the file at `/home/agreen/.config/gcloud/legacy_credentials/a.f.green1@gmail.com/.boto` with the one on my laptop shows that parallel_thread_count=5 on the server, while it is unset on the laptop. Removing the parallel_thread_count line from .boto doesn't change the behaviour for me on the server.

The comment I made about 48 threads was simply an assumption that gsutil would use all available threads by default, sorry about that.

**Comment by thobrla** [comment]
Are there any credentials present in ~/.gsutil/credstore?

Could you send a debug log of the failing operation to gs-team@google.com?

You can collect one like so:
`gsutil -D -m cp -r gs://some/path/* ./some/other/path/`

**Comment by thobrla** [comment]
After discussing this offline with the user, it turns out this started happening after the process starvation fix at https://github.com/GoogleCloudPlatform/gsutil/commit/9133250efe8947e99a82ea360d40e6e02dd8f30b was released in gsutil 4.14.  Prior to this change, only ~10 threads were concurrently trying to access the credential lock file, but with unstarved processes that number climbed to ~120 threads.

The solution was to reduce the number of processes and threads from the defaults by setting parallel_process_count and parallel_thread_count in the .boto configuration file.

**Comment by thobrla** [comment]
FYI the default amount of threads on Linux systems has been reduced by a factor of 2 starting with https://github.com/GoogleCloudPlatform/gsutil/commit/e3ab76d8f97ae9f2233a0c51148a234a0869273c
